plugins/node/opentelemetry-plugin-koa/test/koa.test.ts

```diff
@@ -0,0 +1,217 @@
+import { NoopLogger } from '@opentelemetry/core';
+import { NodeTracerProvider } from '@opentelemetry/node';
+import { AsyncHooksContextManager } from '@opentelemetry/context-async-hooks';
+import {
+  InMemorySpanExporter,
+  SimpleSpanProcessor,
+} from '@opentelemetry/tracing';
+import * as assert from 'assert';
+import * as koa from 'koa';
+import * as http from 'http';
+import { AddressInfo } from 'net';
+import { plugin } from '../src';
+import { AttributeNames, KoaLayerType, KoaComponentName } from '../src/types';
+
+const httpRequest = {
+  get: (options: http.ClientRequestArgs | string) => {
+    return new Promise((resolve, reject) => {
+      return http.get(options, resp => {
+        let data = '';
+        resp.on('data', chunk => {
+          data += chunk;
+        });
+        resp.on('end', () => {
+          resolve(data);
+        });
+        resp.on('error', err => {
+          reject(err);
+        });
+      });
+    });
+  },
+};
+
+describe('Koa Plugin - Core Tests', () => {
+  const logger = new NoopLogger();
+  const provider = new NodeTracerProvider();
+  const memoryExporter = new InMemorySpanExporter();
+  const spanProcessor = new SimpleSpanProcessor(memoryExporter);
+  provider.addSpanProcessor(spanProcessor);
+  const tracer = provider.getTracer('default');
+  let contextManager: AsyncHooksContextManager;
+
+  before(() => {
+    plugin.enable(koa, provider, logger);
+  });
+
+  beforeEach(() => {
+    contextManager = new AsyncHooksContextManager();
+    context.setGlobalContextManager(contextManager.enable());
+  });
+
+  afterEach(() => {
+    memoryExporter.reset();
+    context.disable();
+  });
+
+  const simpleResponse: koa.Middleware = (ctx, next) => {
+    ctx.body = 'test';
+    return next();
+  };
+
+  const customMiddleware: koa.Middleware = (ctx, next) => {
+    for (let i = 0; i < 1000000; i++) {
+      continue;
+    }
+    return next();
+  };
+
+  const asyncMiddleware: koa.Middleware = async (ctx, next) => {
+    const start = Date.now();
+    await next();
+    const ms = Date.now() - start;
+    console.log(`${ctx.method} ${ctx.url} - ${ms}ms`);
+  };
+
+  describe('Instrumenting core middleware calls', () => {
+    it('should create a child span for middlewares', async () => {
+      const rootSpan = tracer.startSpan('rootSpan');
+      const app = new koa();
+      app.use((ctx, next) => tracer.withSpan(rootSpan, next));
+      app.use(customMiddleware);
+      app.use(simpleResponse);
+
+      app.use((ctx, next) => {
+        ctx.body = 'anonymous';
+        return next();
+      });
+
+      const server = http.createServer(app.callback());
+      await new Promise(resolve => server.listen(0, resolve));
+      const port = (server.address() as AddressInfo).port;
+      assert.strictEqual(memoryExporter.getFinishedSpans().length, 0);
+
+      await tracer.withSpan(rootSpan, async () => {
+        await httpRequest.get(`http://localhost:${port}`);
+        rootSpan.end();
+
```

Comment:
nit: add test to check for number of finished spans that's expected? (same applies to below test cases)

Reply:
Done! Thanks